File: containers/spike-wm/keeper.py
#!/usr/bin/env python3
"""Persistent dummy pointer ("keeper"): keeps the compositor's seat POINTER
capability alive across Sunshine's device churn, so gamescope's input-thread
wl_pointer is never released (the release/recreate cycle loses wl_pointer.enter
forever -> dead mouse in nested gamescope under labwc). Sends no events."""
import fcntl
import os
import signal
import struct
import time

UI_DEV_CREATE = 0x5501
UI_DEV_DESTROY = 0x5502
UI_SET_EVBIT = 0x40045564
UI_SET_KEYBIT = 0x40045565
UI_SET_RELBIT = 0x40045566

fd = os.open("/dev/uinput", os.O_WRONLY | os.O_NONBLOCK)
fcntl.ioctl(fd, UI_SET_EVBIT, 1)          # EV_KEY
fcntl.ioctl(fd, UI_SET_KEYBIT, 0x110)     # BTN_LEFT
fcntl.ioctl(fd, UI_SET_EVBIT, 2)          # EV_REL
fcntl.ioctl(fd, UI_SET_RELBIT, 0)         # REL_X
fcntl.ioctl(fd, UI_SET_RELBIT, 1)         # REL_Y
dev = struct.pack("80sHHHHi", b"podstage passthrough keeper", 3, 0x1234, 0x5678, 1, 0) + b"\0" * 1024
os.write(fd, dev)
fcntl.ioctl(fd, UI_DEV_CREATE)
print("keeper up", flush=True)

running = True
def stop(*_):
    global running
    running = False
signal.signal(signal.SIGTERM, stop)
signal.signal(signal.SIGINT, stop)
while running:
    time.sleep(5)
fcntl.ioctl(fd, UI_DEV_DESTROY)
os.close(fd)

File: containers/runtime/keeper.c
/* podstage keeper — a persistent, silent uinput pointer device.
 *
 * gamescope's (3.16) Wayland-backend input thread releases its wl_pointer
 * whenever the seat's POINTER capability drops — and Sunshine's virtual
 * devices come and go with every stream, so that happens constantly. The
 * object recreated on the next attach never receives another
 * wl_pointer.enter (wlroots does not replay enter to late-bound resources),
 * after which absolute mouse input is dead for the rest of the session.
 *
 * Keeping one pointer-capable device around for the whole session means the
 * capability never drops and the input thread's wl_pointer survives. The
 * device name matches the host udev rule ("*passthrough*" -> streaming seat
 * + owner chown), and it never emits a single event.
 */
#include <fcntl.h>
#include <linux/uinput.h>
#include <stdio.h>
#include <string.h>
#include <sys/ioctl.h>
#include <unistd.h>

int main(void) {
    int fd = open("/dev/uinput", O_WRONLY | O_NONBLOCK);
    if (fd < 0) {
        perror("keeper: /dev/uinput");
        return 1;
    }
    ioctl(fd, UI_SET_EVBIT, EV_KEY);
    ioctl(fd, UI_SET_KEYBIT, BTN_LEFT);
    ioctl(fd, UI_SET_EVBIT, EV_REL);
    ioctl(fd, UI_SET_RELBIT, REL_X);
    ioctl(fd, UI_SET_RELBIT, REL_Y);

    struct uinput_user_dev dev;
    memset(&dev, 0, sizeof dev);
    strncpy(dev.name, "podstage passthrough keeper", UINPUT_MAX_NAME_SIZE - 1);
    dev.id.bustype = BUS_VIRTUAL;
    dev.id.vendor = 0x1d50;   /* openmoko community space, no real product */
    dev.id.product = 0x0001;
    dev.id.version = 1;
    if (write(fd, &dev, sizeof dev) != (ssize_t)sizeof dev) {
        perror("keeper: write uinput_user_dev");
        return 1;
    }
    if (ioctl(fd, UI_DEV_CREATE) < 0) {
        perror("keeper: UI_DEV_CREATE");
        return 1;
    }
    for (;;)
        pause();   /* the device dies with the process (container stop) */
}
